### PR TITLE
Reorganize the meaning of engine axes.

### PR DESCRIPTION
**Pull request by reduz**
Until not too long ago, Godot did not really have any concept of *forward*, *back*, etc. As the push to make the engine easier to use grew, meanings for each axis unit vector were added to the engine.

The problem, however, is that Godot is an engine that uses the [right handed coordinate system](https://en.wikipedia.org/wiki/Right-hand_rule) with Y as "up axis". Unlike engines that use the Left Hand Rule (Unity, Unreal), this causes a conflict between what the camera considers as *forward* (Z-) and what imported 3D models end up considering as *forward* when converted to this coordinate system (Z+). As such, the use of methods such as *look_at()* or constants such as **Vector3.FORWARD** do not yield the expected results, as the meaning of foward differs depending on the context. 

image

Again, this is not a problem in engines such as Unity due to the chosen coordinate system (which followed the Direct3D standard), where *forward* is the same for both cameras and models, but it is an explicit problem for Godot due to its chosen coordinate system (which followed the OpenGL standard).

As it is obviously more than a decade too late to change the coordinate system in Godot, this PR intends to ameliorate confusion for Godot users, by explicitly separating the concept of *camera forward* and *model front*, and deprecating unique constants that can be ambiguous but conflicting in meaning.

Additionally, this PR ensures that engine axes are renamed to things that make more sense, as now `Vector3::FORWARD` is `Vector3::CAMERA_FORWARD` (pointing to Z-, which confused many users). And a `Vector3::FRONT` is added that is Z+.

This PR also requires #76052 being merged to ensure everything is fully consistent again.

### CODE REVIEW

**Review by QbieShay — CHANGES_REQUESTED** [review]
LGTM. Since the inconsistency is specifically between what is a camera and what isn't, i think the extra parameter should be called `from_camera` (the "phrase it creates is look at from camera or look at not from camera). This way users don't need to have the extra knowledge of knowing that camera use -z, they can just answer the "from camera" question.

**Comment by reduz** [comment]
@QbieShay I am not sure I get which line you refer to, I think you mean in Basis?

File: scene/3d/node_3d.cpp [hunk 1]
```diff
@@ -820,12 +821,48 @@ void Node3D::look_at_from_position(const Vector3 &p_pos, const Vector3 &p_target
...
+	Transform3D lookat = Transform3D(Basis::looking_at(p_target - p_pos, p_up), p_pos);
+	Vector3 original_scale = get_scale();
+	set_global_transform(lookat);
+	set_scale(original_scale);
+	WARN_DEPRECATED_MSG("This function is deprecated and will eventually be removed. Please use camera_look_at() instead, or model_look_at() functions depending on the right context.");
+}
+#endif
+
+void Node3D::camera_look_at(const Vector3 &p_target, const Vector3 &p_up) {
+	ERR_FAIL_COND_MSG(!is_inside_tree(), "Node not inside tree. Use look_at_from_position() instead.");
```
Comment (QbieShay) [inline scene/3d/node_3d.cpp:833]: Look at from position is deprecated

**Review by QbieShay — CHANGES_REQUESTED** [review]

**Comment by QbieShay** [comment]
> @QbieShay I am not sure I get which line you refer to, I think you mean in Basis?

Yep sorry, i thought i commented directly on the code.

**Review by TokageItLab — APPROVED** [review]
I approve this PR as one of the people who worked on the animation and glTF pipeline workflow. I almost agree with reduz.

There has been a lot of discussion about having two directions, but I believe we cannot to avoid this way during Godot 4.x, so we need to put an end to the -Z/+Z Forward issue today. I can't have idea what kind of requests we will get in Godot 5 until after this is merged. 

After this PR is merged, a more detailed article about pipeline is needed on the document or blogpost.

Also, a same option needs to be added to #72842 (That PR doesn't pass the test as there is no constant link to the class-reference yet).

File: doc/classes/Node3D.xml [hunk 2]
```diff
@@ -20,6 +20,26 @@
...
+			<return type="void" />
+			<param index="0" name="target" type="Vector3" />
+			<param index="1" name="up" type="Vector3" default="Vector3(0, 1, 0)" />
+			<description>
+				Rotates the node so that the [constant Vector3.CAMERA_FORWARD] axis (-Z) points toward the [param target] position. Use this method when you want to point a Camera3D node towards a specific positions. For 3D models, [method model_look_at] will properly orient the front in the desired direction.
+				The local up axis (+Y) points as close to the [param up] vector as possible while staying perpendicular to the local forward axis. The resulting transform is orthogonal, and the scale is preserved. Non-uniform scaling may not work correctly.
+				The [param target] position cannot be the same as the node's position, the [param up] vector cannot be zero, and the direction from the node's position to the [param target] vector cannot be parallel to the [param up] vector.
+				Operations take place in global space, which means that the node must be in the scene tree.
+				This method is deprecated, please use [method model_look_at] or [method camera_look_at_from_position] depending on your needs.
```
Comment (YuriSizov) [inline doc/classes/Node3D.xml:32]: This note should probably not be here, since this is a new method.

File: core/variant/variant_call.cpp [hunk 3]
```diff
@@ -2522,12 +2522,40 @@ static void _register_variant_builtin_methods() {
...
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "VIEW_RIGHT", Vector3(1, 0, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "VIEW_TOP", Vector3(0, 1, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "VIEW_BOTTOM", Vector3(0, -1, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "VIEW_FRONT", Vector3(0, 0, 1));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "VIEW_BACK", Vector3(0, 0, -1));
+
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "NORTH", Vector3(0, 0, -1));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "SOUTH", Vector3(0, 0, 1));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "EAST", Vector3(1, 0, 0));
+	_VariantCall::add_variant_constant(Variant::VECTOR3, "WEST", Vector3(-1, 0, 0));
```
Comment (clayjohn) [inline core/variant/variant_call.cpp:2558]: Would a ``GLOBAL_`` prefix make more sense here? ``Like GLOBAL_FORWARD``, ``GLOBAL_LEFT`` etc.

I wouldn't think to use NORTH, SOUTH etc. if I was looking for constants in global space.

File: scene/3d/node_3d.cpp (hunk 1 above)
Comment (clayjohn) [reply scene/3d/node_3d.cpp:833]:
```suggestion
	ERR_FAIL_COND_MSG(!is_inside_tree(), "Node not inside tree. Use camera_look_at_from_position() instead.");
```

File: scene/3d/node_3d.cpp [hunk 4]
```diff
@@ -820,12 +821,48 @@ void Node3D::look_at_from_position(const Vector3 &p_pos, const Vector3 &p_target
...
+	ERR_FAIL_COND_MSG(p_up.cross(p_target - p_pos).is_zero_approx(), "Up vector and direction between node origin and target are aligned, camera_look_at() failed.");
+
 	Transform3D lookat = Transform3D(Basis::looking_at(p_target - p_pos, p_up), p_pos);
 	Vector3 original_scale = get_scale();
 	set_global_transform(lookat);
 	set_scale(original_scale);
 }
 
+void Node3D::model_look_at(const Vector3 &p_target, const Vector3 &p_up) {
+	ERR_FAIL_COND_MSG(!is_inside_tree(), "Node not inside tree. Use look_at_from_position() instead.");
```
Comment (clayjohn) [inline scene/3d/node_3d.cpp:850]:
```suggestion
	ERR_FAIL_COND_MSG(!is_inside_tree(), "Node not inside tree. Use model_look_at_from_position() instead.");
```

File: core/variant/variant_call.cpp (hunk 3 above)
Comment (reduz) [reply core/variant/variant_call.cpp:2558]: alright

File: doc/classes/Node3D.xml [hunk 5]
```diff
@@ -117,15 +136,37 @@
 				The local up axis (+Y) points as close to the [param up] vector as possible while staying perpendicular to the local forward axis. The resulting transform is orthogonal, and the scale is preserved. Non-uniform scaling may not work correctly.
 				The [param target] position cannot be the same as the node's position, the [param up] vector cannot be zero, and the direction from the node's position to the [param target] vector cannot be parallel to the [param up] vector.
 				Operations take place in global space, which means that the node must be in the scene tree.
+				This method is deprecated, please use [method model_look_at] or [method camera_look_at_from_position] depending on your needs.
```
Comment (lyuma) [inline doc/classes/Node3D.xml:139]:
```suggestion
				This method is deprecated, please use [method model_look_at] or [method camera_look_at] depending on your needs.
```

**Comment by timothyqiu** [comment]
> Again, this is not a problem in engines such as Unity due to the chosen coordinate system (which followed the Direct3D standard), where _forward_ is the same for both cameras and models, but it is an explicit problem for Godot due to its chosen coordinate system (which followed the OpenGL standard).

My nitpick is that coordinate system choice should not be the one to be blamed here.

The root cause is that we imported glTF models as-is. glTF stores the model +Z forward, -X right, while Godot uses -Z forward, +X right.

A model facing forward in Blender is exported as facing +Z in glTF. But Godot keeps the imported model facing +Z in TSCN, which is backward facing.

* So the front view is viewing the model's butt.
* So the `look_at()` function don't work for models.

The ideal solution is to convert the coordinate system when importing. But this would also require large modifications to systems like animation and is probably compat breaking. So it's too late to do things the correct way.

So we decided to keep the imported model as-is last night, and:

* the front view is viewing the model's butt: flip the front view and back view.
* the `look_at()` function don't work for models: add an extra argument to look using the +Z axis.

I think we must admit that this is a workaround for a problem that we can't afford to fix correctly.

**Comment by reduz** [comment]
@timothyqiu I understand this point of view, but if you look at virtually any other software that uses this coordinate system, it uses imported the models the same way, so besides the fact that it's not a great idea at this point of time to innovate by importing the models flipped and flipping all the camera management in the editor, its also not a good idea IMO to create yet another different coordinate system for models (which would be something like RHS, Y-up, flipped 180 in Y).

**Comment by Iniquitatis** [comment]
> I think we must admit that this is a workaround for a problem that we can't afford to fix correctly.

If this is all about models, can't there be an _option_ (maybe recommended or enabled by default until the next major release) to simply flip models on import? This way there would be no need to introduce more band-aids to the coordinate system manipulation that certainly won't make the engine easier to learn/use for newcomers, while the existing users could gradually move their projects towards this model flipped approach. Switching brain between two modes isn't something that I'd call errorproof.

**Comment by TokageItLab** [comment]
@Iniquitatis In Godot 4 which uses the right-hand coordinate system, the model and camera have different forwards, which is the canonical way to do it, so I believe there is no chance for that.

If you want the model and the camera forward to be the same, and you believe that is the right thing to do, then you need the proposal of adopting a left-handed coordinate system in Godot 5.

For those who really want to hack reproduce the left-hand coordinate system by rotating the imported model to -Z, maybe an only possible work-around is to make https://github.com/godotengine/godot/pull/72753 an add-on and only enable the add-on to rotate the model by default for those who have enabled it.

File: doc/classes/Node3D.xml [hunk 6]
```diff
@@ -117,15 +136,37 @@
...
 				Moves the node to the specified [param position], and then rotates the node to point toward the [param target] as per [method look_at]. Operations take place in global space.
+				This method is deprecated, please use [method model_look_at_from_position] or [method camera_look_at_from_position] depending on your needs.
+			</description>
+		</method>
+		<method name="model_look_at">
+			<return type="void" />
+			<param index="0" name="target" type="Vector3" />
+			<param index="1" name="up" type="Vector3" default="Vector3(0, 1, 0)" />
+			<description>
+				Rotates the node so that the [constant Vector3.MODEL_FRONT] axis (+Z) points toward the [param target] position. Use this method when you want to point a 3D model towards a specific position. For 3D cameras ([Camera3D]) use [method camera_look_at].
```
Comment (capnm) [inline doc/classes/Node3D.xml:157]: Vector3.FRONT

**Comment by QbieShay** [comment]
@Iniquitatis while we haven't entirely reached a consensus on what a future solution may look like, we took long time to discuss this in rocket chat and rotating the imported models cannot be done at this stage of Godot4, because it requires changes to everything (meshes, animations, retargeting). If it's made by default it will break everything made so far with Godot4 (we're at stable phase) and if we make it optional everything has to support the multiple option and it's going to make the code way too complicated. Changing look_at and reverting the front view in editor is the only compromise that's functional at this point in time.

**Comment by aaronfranke** [comment]
@Iniquitatis See this PR I opened #72753. This adds an import setting to rotate by 180 degrees. It is completely optional, off by default, so it does not break compatibility. It also does not conflict with this PR so we can have both.

**Comment by TokageItLab** [comment]
Superseded by #76082.

**Pull request closed without merging**